File: apps/web/src/store/game-store.ts
'use client';

import { create } from 'zustand';
import type { Board, Move, Player, GameStatus, Difficulty, GameMode } from '@raichu/shared-types';
import {
  createInitialBoard,
  generateAllMoves,
  generateMovesForPiece,
  applyMove,
  getGameStatus,
  encodeBoard,
} from '@raichu/game-engine';
import { analytics } from '../lib/analytics';
import { computeGameSummary } from '../lib/play-style';
import { findBestMoveOffThread } from '../lib/ai-worker-client';

let botSearchId = 0;

interface GameStore {
  // Game state
  board: Board;
  currentPlayer: Player;
  status: GameStatus;
  moveHistory: Move[];
  boardHistory: Board[];
  lastMove: Move | null;

  // Play-style tracking (offline games never reach the `moves` table, so the
  // timing data has to be captured here or it is lost)
  gameStartedAt: number;
  moveTimestamps: number[];

  // Draw detection
  positionCounts: Record<string, number>;  // "boardEncoded:player" → occurrences
  halfMovesSinceProgress: number;          // resets on capture or promotion
  drawReason: string | null;

  // Game config
  gameMode: GameMode;
  difficulty: Difficulty;
  playerColor: Player;

  // UI state
  selectedPiece: { row: number; col: number } | null;
  legalMoves: Move[];
  isThinking: boolean;

  // Actions
  selectPiece: (row: number, col: number) => void;
  makeMove: (move: Move) => void;
  clearSelection: () => void;
  newGame: (config: { mode: GameMode; difficulty: Difficulty; playerColor: Player }) => void;
  restart: () => void;
  requestBotMove: () => Promise<void>;
}

export const useGameStore = create<GameStore>((set, get) => ({
  // Initial state
  board: createInitialBoard(),
  currentPlayer: 'white',
  status: 'playing',
  moveHistory: [],
  boardHistory: [createInitialBoard()],
  lastMove: null,
  gameStartedAt: Date.now(),
  moveTimestamps: [],
  positionCounts: { [`${encodeBoard(createInitialBoard())}:white`]: 1 },
  halfMovesSinceProgress: 0,
  drawReason: null,
  // Default to a bot game: the home page's primary CTA says "Play vs AI" and
  // links here, so pvp meant a first-time visitor silently played themselves.
  // Easy (400ms search) rather than medium (800ms) because every bot reply
  // blocks the main thread, and a slow first move reads as the page being broken.
  // playerColor must stay 'white' — the initial state has no equivalent of
  // newGame's bot-first kick-off, so a black default would never move.
  gameMode: 'bot',
  difficulty: 'easy',
  playerColor: 'white',
  selectedPiece: null,
  legalMoves: [],
  isThinking: false,

  selectPiece: (row: number, col: number) => {
    const { board, currentPlayer, gameMode, playerColor, isThinking, status } = get();
    if (status !== 'playing' || isThinking) return;

    // In bot mode, only allow selecting own pieces
    if (gameMode === 'bot' && currentPlayer !== playerColor) return;

    const moves = generateMovesForPiece(board, row, col);
    // Filter to only show moves for the current player's pieces
    const playerMoves = moves.filter(() => {
      const cell = board[row][col];
      if (cell === '.') return false;
      const isWhite = cell === 'w' || cell === 'W' || cell === '@';
      return currentPlayer === 'white' ? isWhite : !isWhite;
    });

    if (playerMoves.length > 0) {
      set({ selectedPiece: { row, col }, legalMoves: playerMoves });
    } else {
      set({ selectedPiece: null, legalMoves: [] });
    }
  },

  makeMove: (move: Move) => {
    const {
      board, currentPlayer, moveHistory, boardHistory, gameMode, playerColor,
      positionCounts, halfMovesSinceProgress, gameStartedAt, moveTimestamps,
    } = get();

    const movedAt = Date.now();
    const newMoveTimestamps = [...moveTimestamps, movedAt];

    const newBoard    = applyMove(board, move);
    const nextPlayer: Player = currentPlayer === 'white' ? 'black' : 'white';

    // Base status from engine (piece count + no-legal-moves)
    let finalStatus: GameStatus = getGameStatus(newBoard, nextPlayer);

    // ── Draw detection ────────────────────────────────────────────────────

    // 1) Track half-moves without capture or promotion (50-move rule)
    const newHalfMoves = (move.captured || move.promotion) ? 0 : halfMovesSinceProgress + 1;

    // 2) Threefold repetition — position key = board encoding + side to move
    const posKey = `${encodeBoard(newBoard)}:${nextPlayer}`;
    const newPosCount = (positionCounts[posKey] ?? 0) + 1;
    const newPositionCounts = { ...positionCounts, [posKey]: newPosCount };

    let drawReason: string | null = null;
    if (finalStatus === 'playing') {
      if (newPosCount >= 3) {
        finalStatus = 'draw';
        drawReason  = 'Threefold repetition';
      } else if (newHalfMoves >= 100) {
        // 50 full moves without capture or promotion
        finalStatus = 'draw';
        drawReason  = '50-move rule';
      }
    }

    // ─────────────────────────────────────────────────────────────────────

    set({
      board: newBoard,
      currentPlayer: nextPlayer,
      status: finalStatus,
      moveHistory: [...moveHistory, move],
      moveTimestamps: newMoveTimestamps,
      boardHistory: [...boardHistory, newBoard],
      lastMove: move,
      positionCounts: newPositionCounts,
      halfMovesSinceProgress: newHalfMoves,
      drawReason,
      selectedPiece: null,
      legalMoves: [],
    });

    if (finalStatus !== 'playing') {
      const newMoveHistory = [...moveHistory, move];

      analytics.gameEnded({
        mode:        gameMode,
        status:      finalStatus,
        moveCount:   newMoveHistory.length,
        duration_ms: movedAt - gameStartedAt,
      });

      // One summary per game rather than an event per move: same analytical
      // value at roughly 1/40th the row count.
      analytics.gameSummary(
        computeGameSummary({
          moveHistory: newMoveHistory,
          moveTimestamps: newMoveTimestamps,
          startedAt: gameStartedAt,
          endedAt: movedAt,
          playerColor,
          mode: gameMode,
          difficulty: get().difficulty,
          status: finalStatus,
        }),
      );
    }

    if (gameMode === 'bot' && finalStatus === 'playing' && nextPlayer !== playerColor) {
      setTimeout(() => get().requestBotMove(), 300);
    }
  },

  clearSelection: () => {
    set({ selectedPiece: null, legalMoves: [] });
  },

  newGame: (config) => {
    botSearchId += 1;
    const initial = createInitialBoard();
    analytics.gameStarted({
      mode:        config.mode,
      difficulty:  config.difficulty,
      playerColor: config.playerColor,
    });
    set({
      board: initial,
      currentPlayer: 'white',
      status: 'playing',
      moveHistory: [],
      gameStartedAt: Date.now(),
      moveTimestamps: [],
      boardHistory: [initial],
      lastMove: null,
      positionCounts: { [`${encodeBoard(initial)}:white`]: 1 },
      halfMovesSinceProgress: 0,
      drawReason: null,
      gameMode: config.mode,
      difficulty: config.difficulty,
      playerColor: config.playerColor,
      selectedPiece: null,
      legalMoves: [],
      isThinking: false,
    });

    // If bot goes first
    if (config.mode === 'bot' && config.playerColor !== 'white') {
      setTimeout(() => get().requestBotMove(), 500);
    }
  },

  restart: () => {
    const { gameMode, difficulty, playerColor } = get();
    get().newGame({ mode: gameMode, difficulty, playerColor });
  },

  requestBotMove: async () => {
    const initial = get();
    if (
      initial.status !== 'playing' ||
      initial.gameMode !== 'bot' ||
      initial.currentPlayer === initial.playerColor ||
      initial.isThinking
    ) return;

    const requestId = ++botSearchId;

    set({ isThinking: true });

    // Yield to React so the "Thinking…" state renders before the synchronous
    // AI search blocks the main thread.
    await new Promise<void>(resolve => setTimeout(resolve, 50));

    try {
      const { board, currentPlayer, difficulty, status: currentStatus } = get();
      if (currentStatus !== 'playing') return;

      const position = encodeBoard(board);
      const result = await findBestMoveOffThread(board, currentPlayer, difficulty);

      const latest = get();
      if (
        requestId !== botSearchId ||
        latest.status !== 'playing' ||
        latest.currentPlayer !== currentPlayer ||
        encodeBoard(latest.board) !== position
      ) return;

      if (result.move) {
        get().makeMove(result.move);
      }
    } catch {
      // Last-resort: pick any legal move so the game never deadlocks
      const { board, currentPlayer } = get();
      if (requestId === botSearchId) {
        const moves = generateAllMoves(board, currentPlayer);
        if (moves.length > 0) {
          get().makeMove(moves[0]);
        }
      }
    } finally {
      if (requestId === botSearchId) set({ isThinking: false });
    }
  },
}));

File: apps/Ios-app/Raichu/Raichu/Stores/GameStore.swift
// GameStore.swift
// Raichu
// Offline game state: PvP and bot games using the local JS engine.

import SwiftUI
import Combine

@MainActor
final class GameStore: ObservableObject {
    // Board & game state
    @Published var board: [[String]] = []
    @Published var currentPlayer: String = "white"
    @Published var status: String = "playing"
    @Published var moveHistory: [Move] = []
    @Published var boardHistory: [[[String]]] = []
    @Published var lastMove: Move? = nil
    @Published var isThinking: Bool = false

    // Selection UI state
    @Published var selectedPiece: Position? = nil
    @Published var legalMoves: [Move] = []

    // Draw detection
    @Published var positionCounts: [String: Int] = [:]
    @Published var halfMovesSinceProgress: Int = 0
    @Published var drawReason: String? = nil

    // Config
    @Published var gameMode: String = "bot"        // pvp | bot
    @Published var difficulty: String = "easy"     // easy | medium | hard
    @Published var playerColor: String = "white"   // which color the human plays (bot mode)

    init() {
        board = RaichuEngine.shared.createInitialBoard()
    }

    // MARK: - Selection

    func selectPiece(at pos: Position) {
        let cell = board[pos.row][pos.col]

        // If tapping a legal move destination
        if let selected = selectedPiece,
           let move = legalMoves.first(where: { $0.to == pos }) {
            _ = selected  // consumed
            makeMove(move)
            return
        }

        // Tapping own piece
        if cell.pieceColor?.rawValue == currentPlayer {
            selectedPiece = pos
            legalMoves = RaichuEngine.shared.generateMovesForPiece(
                board: board, row: pos.row, col: pos.col
            )
        } else {
            clearSelection()
        }
    }

    func clearSelection() {
        selectedPiece = nil
        legalMoves = []
    }

    // MARK: - Move Execution

    func makeMove(_ move: Move) {
        let newBoard = RaichuEngine.shared.applyMove(board: board, move: move)
        let nextPlayer = currentPlayer == "white" ? "black" : "white"

        // Draw detection: half-move clock
        let isCapture = move.captured != nil
        let isPromotion = move.promotion != nil
        if isCapture || isPromotion {
            halfMovesSinceProgress = 0
        } else {
            halfMovesSinceProgress += 1
        }

        // Threefold repetition tracking
        let key = RaichuEngine.shared.encodeBoard(newBoard) + ":\(nextPlayer)"
        let count = (positionCounts[key] ?? 0) + 1
        positionCounts[key] = count

        // Update board history for replay
        boardHistory.append(board)
        moveHistory.append(move)

        // Apply state
        board = newBoard
        currentPlayer = nextPlayer
        lastMove = move
        clearSelection()

        // Check game status
        let gameStatus = RaichuEngine.shared.getGameStatus(board: newBoard, nextPlayer: nextPlayer)

        if count >= 3 {
            status = "draw"
            drawReason = "Threefold repetition"
        } else if halfMovesSinceProgress >= 100 {
            status = "draw"
            drawReason = "50-move rule"
        } else {
            status = gameStatus
        }

        // Trigger bot if applicable
        if gameMode == "bot" && status == "playing" && currentPlayer != playerColor {
            Task {
                try? await Task.sleep(nanoseconds: 300_000_000) // 0.3s delay for UI
                await requestBotMove()
            }
        }
    }

    // MARK: - Bot

    func requestBotMove() async {
        guard status == "playing", gameMode == "bot", currentPlayer != playerColor else { return }
        isThinking = true
        defer { isThinking = false }

        // Yield for UI update
        try? await Task.sleep(nanoseconds: 50_000_000) // 50ms

        let botMove = RaichuEngine.shared.findBestMove(
            board: board, player: currentPlayer, difficulty: difficulty
        )

        if let move = botMove {
            makeMove(move)
        } else if let randomMove = RaichuEngine.shared.generateAllMoves(board: board, player: currentPlayer).randomElement() {
            makeMove(randomMove)
        }
    }

    // MARK: - Game Control

    func newGame(mode: String = "bot", difficulty: String = "easy", playerColor: String = "white") {
        self.gameMode = mode
        self.difficulty = difficulty
        self.playerColor = playerColor

        board = RaichuEngine.shared.createInitialBoard()
        currentPlayer = "white"
        status = "playing"
        moveHistory = []
        boardHistory = []
        lastMove = nil
        isThinking = false
        positionCounts = [:]
        halfMovesSinceProgress = 0
        drawReason = nil
        clearSelection()

        // If bot game and bot goes first
        if mode == "bot" && playerColor == "black" {
            Task {
                try? await Task.sleep(nanoseconds: 500_000_000)
                await requestBotMove()
            }
        }
    }

    func restart() {
        newGame(mode: gameMode, difficulty: difficulty, playerColor: playerColor)
    }

    // MARK: - Captured Pieces

    var capturedByWhite: [String] {
        moveHistory.compactMap { $0.captured?.piece }.filter { $0.isBlackPiece }
    }

    var capturedByBlack: [String] {
        moveHistory.compactMap { $0.captured?.piece }.filter { $0.isWhitePiece }
    }
}
